DM-2432 Removing sensor-library chapter. Ranamed device-management-library. (#1181)
SoftwareAG/c8y-docs: content/reference/device-integrator-library-bundle/relay.md

```
|Name|Type|Mandatory|Description|
|:---|:---|:--------|:----------|
|c8y_RelayArray|array|Yes|Array of strings of relays states.|

#### Setting multiple relays states

Devices that support changing their relays positions remotely may add the `c8y_RelayArray` operation to `c8y_SupportedOperations`. Then users can request an update to the relay position in the [Relay array control widget](/users-guide/cockpit#relay-array-control). This creates an operation of fragment type `c8y_RelayArray` for the device.

The operation representation is the same as the inventory representation:

```json
"c8y_RelayArray" : [
"OPEN",
"CLOSED",
"CLOSED",
"OPEN"
]
```

On receiving the operation the device is expected to perform the following actions:

1. Set operation status to EXECUTING
2. Set the relays states to the respective values.
3. Set new relays states to its managed object
4. Set operation status to SUCCESSFUL

**SmartREST example**

{{< product-c8y-iot >}} provides the 519 static response template:

1. The device receives the command via the 519 static response template<br>
   `519,OPEN,CLOSED,CLOSED,OPEN`
2. The device sets the operation status to EXECUTING<br>
   `501,c8y_RelayArray`

> 3. The device sets its relays states.
> 4. The device confirms successful execution by setting the operation status to SUCCESSFUL<br>
>    `503,c8y_RelayArray`

Reviewer: ```suggestion
3. The device confirms successful execution by setting the operation status to SUCCESSFUL and sets the relay array state with the implicit parameters in the 503 static template<br>
   `503,c8y_RelayArray,OPEN,CLOSED,CLOSED,OPEN`
```

---
weight: 85
title: Relay
layout: redirect

> ---

Reviewer: This whole file needs to be aligned with the rest of this documentation. This doc explains ho device integrations can implement certain functionalities into their devices. The focus should be on the fuctionality of monitoring and managing a single or multiple two-position switches on a device.
There are two separate relay relay widgets in Cockpit: Relay control (for a single switch) and Relay array control (for multiple switches) for each of these we need a chapter for how a device can announce the status to the platform and a section how controlling them works. All of those with JSON data model examples of the fragments a device should put into its MO and the operation, examples with SmartREST ad descriptions of steps expected from the device.

Reviewer: If you look at other segments in this doc they are structured in a very particular way that is specifically around explaining how to integrate certain functionalities into a device and what the platform expects from a device. A good example for this chapter is the network section. This is a much more complex operation but ultimately a very similar concept with announcing the status and the request chang operation which are always separate flows. Here is the structure that I would have expected:

### Relays
Text is ok example and link should be moved to the respective sub section

#### Single Relay
Text similar to: "To manage and monitor a single relay Cumulocity offers the Relay control widget"

##### Relay status
Text similar to: "Devices may announce their relay status by updating the c8y_Relay Fragment in their managed object"
<JSON example>
<SmartREST example>

##### Setting relay position
Text similar to: "Devices that support changing the relay position remotely may add the c8y_Relay operation to c8y_SupportedOperations. Then Users can request an update to the relay position in the Relay control widget. This creates an operation of fragment type c8y_Relay for the device.
<JSON example>
<Steps expected from the device>

1. set op to executing
2. Apply requested relay position
3. update relay status in MO
4. set op to successful

<SmartREST examples of all steps>

#### Multiple relays
<Repeat similar structure>

Author: Ok, tried one more time.

}
```

> For devices with `c8y_Relay` fragment [relay control widget](/users-guide/cockpit#relay-control) can be used.

Reviewer 2: ```suggestion
For devices with the `c8y_Relay` fragment the [Relay control widget](/users-guide/cockpit#relay-control) can be used.
```

}
```

**SmartREST example**

{{< product-c8y-iot >}} provides the 518 static response template:

> 1. Device receives command via 518 static response template <br>

Reviewer 2: ```suggestion
1. The device receives the command via the 518 static response template<br>
```

   `518,OPEN`

> 2. Device sets operation status to EXECUTING <br>

Reviewer 2: ```suggestion
2. The device sets the operation status to EXECUTING<br>
```

   `501,c8y_Relay`

> 3. Device confirms successful execution by setting operation status to SUCCESSFUL <br>

Reviewer 2: ```suggestion
3. The device confirms successful execution by setting the operation status to SUCCESSFUL<br>
```

Author: Probably I will never comprehend where to use "the" in a (the?) programming context.

]
```

> For devices with `c8y_Relay` fragment [relay array control widget](/users-guide/cockpit#relay-array-control) can be used.

Reviewer 2: ```suggestion
For devices with the `c8y_Relay` fragment the [Relay array control widget](/users-guide/cockpit#relay-array-control) can be used.
```

  "OPEN"
]
```

**SmartREST example**

{{< product-c8y-iot >}} provides the 519 static response template:

> 1. Device receives command via 518 static response template <br>

Reviewer 2: ```suggestion
1. The device receives the command via the 519 static response template<br>
```

   `519,OPEN,CLOSED,CLOSED,OPEN`

> 2. Device sets operation status to EXECUTING <br>

Reviewer 2: ```suggestion
2. The device sets the operation status to EXECUTING<br>
```

   `501,c8y_RelayArray`

> 3. Device confirms successful execution by setting operation status to SUCCESSFUL <br>

Reviewer 2: ```suggestion
3. The device confirms successful execution by setting the operation status to SUCCESSFUL<br>
```
